In [2]:
import pandas as pd
from glob import glob
from os import path


In [27]:
def load(indir):
    dfs = []
    for infile in sorted(glob(f"{indir}/*.csv")):
        base = path.splitext(path.basename(infile))[0]
        if base == "index":
            continue
        dfs.append(pd.read_csv(infile, dtype=str, na_filter=False).assign(file=base))
    return pd.concat(dfs)

norm = load("AESI-norm")
norm

sab               code  \
0          ICD10                I74   
1          ICD10              I74.0   
2          ICD10              I74.1   
3          ICD10              I74.2   
4          ICD10              I74.3   
..           ...                ...   
787  SNOMEDCT_US          698247007   
788  SNOMEDCT_US          698247007   
789  SNOMEDCT_US          698247007   
790  SNOMEDCT_US          720448006   
791  SNOMEDCT_US  15964901000119107   

                                                   str       cui  \
0                     Arterial embolism and thrombosis  C0155749   
1           Embolism and thrombosis of abdominal aorta             
2    Embolism and thrombosis of other and unspecifi...             
3    Embolism and thrombosis of arteries of upper e...             
4    Embolism and thrombosis of arteries of lower e...             
..                                                 ...       ...   
787                      Other cardiac dysrhythmia NOS             
788                         Other cardiac dysrhythmias             
789                                 Cardiac arrhythmia  C0003811   
790                             Typical atrial flutter  C3264370   
791                            Atypical atrial flutter  C3264372   

                        concept_name    tags comments  \
0    Arterial embolus and thrombosis  Narrow            
1                                     Narrow            
2                                     Narrow            
3                                     Narrow            
4                                     Narrow            
..                               ...     ...      ...   
787                                   Narrow            
788                                   Narrow            
789               Cardiac Arrhythmia  Narrow            
790           Typical atrial flutter  Narrow            
791          Atypical atrial flutter  Narrow            

                                          file review_author_0  \
0    B_ARTERIALEMBOLISM_AESI_arterial embolism             NaN   
1    B_ARTERIALEMBOLISM_AESI_arterial embolism             NaN   
2    B_ARTERIALEMBOLISM_AESI_arterial embolism             NaN   
3    B_ARTERIALEMBOLISM_AESI_arterial embolism             NaN   
4    B_ARTERIALEMBOLISM_AESI_arterial embolism             NaN   
..                                         ...             ...   
787                  C_ARRH_AESI_Arrythmia all   Joan/Cristina   
788                  C_ARRH_AESI_Arrythmia all   Joan/Cristina   
789                  C_ARRH_AESI_Arrythmia all                   
790                  C_ARRH_AESI_Arrythmia all                   
791                  C_ARRH_AESI_Arrythmia all                   

           review_date_0 review_content_0 review_author_1  \
0                    NaN              NaN             NaN   
1                    NaN              NaN             NaN   
2                    NaN              NaN             NaN   
3                    NaN              NaN             NaN   
4                    NaN              NaN             NaN   
..                   ...              ...             ...   
787  2022-02-01 00:00:00                     Joan Fortuny   
788  2022-02-01 00:00:00                     Joan Fortuny   
789                                          Joan Fortuny   
790                                          Joan Fortuny   
791                                          Joan Fortuny   

           review_date_1 review_content_1 review_author_2  \
0                    NaN              NaN             NaN   
1                    NaN              NaN             NaN   
2                    NaN              NaN             NaN   
3                    NaN              NaN             NaN   
4                    NaN              NaN             NaN   
..                   ...              ...             ...   
787  2022-12-12 00:00:00        no change            sima   
788  2022-12-12 00:00:00    

In [103]:
dedup['sab code dedup_result'.split()].drop_duplicates().dedup_result.value_counts()

EXACT                    1562
BY_CODE_EQUIV             379
NONE                      204
BY_CODE                   174
CHANGE_CODING_SYSTEM        9
NONE_NO_CODING_SYSTEM       9
BY_NAME                     4
CODE_BY_CUI                 1
Name: dedup_result, dtype: int64

In [109]:
dedup_none = dedup[dedup.dedup_result == "NONE"]['cui sab code str'.split()]
dedup_none.sab.value_counts()

MEDCODEID    157
ICD10DA      128
ICPC           6
ICD10          2
SCTSPA         1
ICPC2P         1
Name: sab, dtype: int64

In [110]:
dedup_none[~dedup_none.sab.isin('MEDCODEID ICD10DA'.split())]

cui     sab               code  \
132        ICPC             B83.01   
133        ICPC             B83.02   
134        ICPC             B83.06   
132        ICPC             B83.01   
133        ICPC             B83.02   
134        ICPC             B83.06   
127      SCTSPA  10752381000119100   
3         ICD10             D57.01   
22       ICPC2P                B83   
36        ICD10              I48.9   

                                                   str  
132                                     Hemophilia A/B  
133          Idiopathic thrombocytopenic purpura (ITP)  
134                     Increased coagulation tendency  
132                                     Hemophilia A/B  
133          Idiopathic thrombocytopenic purpura (ITP)  
134                     Increased coagulation tendency  
127                             Fetal thrombocytopenia  
3                   Sickle-cell anaemia without crisis  
22                                                      
36   Atrial fibrillation and atrial flutter, unspec...

In [ ]:
10752381000119100
10752381000119101

In [7]:
norm[['file', 'coding_system']].drop_duplicates().coding_system.value_counts()

ICD9CM         29
ICD10CM        29
ICD10          28
SNOMEDCT_US    28
SCTSPA         27
ICPC2P         25
MEDCODEID      25
RCD2           25
ICD10DA        23
ICPC           17
Free_text      16
MTHICD9        11
SNM             7
RCD             5
SNMI            1
p               1
ICD10/CM        1
e               1
                1
Name: coding_system, dtype: int64

In [10]:
norm[norm.coding_system.isin(["p", "e", ""])][['file', 'coding_system']].drop_duplicates()

file coding_system
227             B_COAGDEF_AESI_Coagulation deficiencies             e
261             B_COAGDEF_AESI_Coagulation deficiencies             p
1573  C_CARDIOMYOPATHY_COV_Cardiomyopathy excl. myo ...

## Coding systems

In [6]:
def coding_systems(indir):
    res = set()
    for infile in sorted(glob(f"{indir}/*.csv")):
        if infile == f"{indir}/index.csv":
            continue
        df = pd.read_csv(infile, dtype=str, na_filter=False)
        for sab in df.coding_system:
            if sab:
                res.add(sab)
    return sorted(res)

for sab in coding_systems("AESI-norm"):
    print(f"- {sab}")

- CVB
- EXCLU
- Free_text
- HCPCS
- HCPT
- HSD_ACCERT
- IC9CM
- ICD-10-CM
- ICD-10-PCS
- ICD-9-CM
- ICD-9CM
- ICD10
- ICD10-PCS
- ICD10/CM
- ICD10CM
- ICD10DA
- ICD9-CM
- ICD9CM
- ICD9CM_HSD
- ICPC
- ICPC2
- ICPC2EENG
- ICPC2P
- MDR
- MEDCODEID
- MTHICD9
- NCMP-NCSP-NCRP
- PROCDA
- RCD
- RCD2
- SCTSPA
- SCTSPA_SNS
- SNM
- SNMI
- SNOMED
- SNOMEDCT_US
- SNOMED_US
- SPA_EXT
- ZA


echo 'select state from case_definitions ' \
| sudo -u postgres psql codemapper \
| head -n-1 | tail -n+3 \
| jq '.codingSystems' \
| grep '^  "' | sed 's/^  "\([^"]*\)",*$/\1/' \
| sort | uniq -c | sort -h

      1 ICPCNOR
      1 KCD5
      1 RCDAE
      1 WHOSPA
      2 CCSR_ICD10PCS
      2 ICPC2ICD10DUT
      2 ICPC2ICD10ENG
      2 LNC
      3 ICPCDUT
      4 SNMI
      9 HCPT
     10 HCPCS
     15 ICD10PCS
     15 ICPC2EDUT
     27 MSH
     65 SNM
     76 MDR
    189 ICD10/CM
    227 RCD
    326 ICPC2EENG
    364 ICPC
    416 SCTSPA
    446 MTHICD9
    501 SNOMEDCT_US
    583 ICPC2P
    607 ICD10
    623 RCD2
    632 ICD10CM
    731 ICD9CM


In [85]:
col_names = {"coding_system": "sab", "term": "str", "code_name": "str", "tag": "tags"}

final_v0 = load("VAC4EU-v0/AESI-final").rename(col_names, axis=1)
final = load("AESI-final").rename(col_names, axis=1)

cols = "sab code".split()
diff = pd.merge(
    final_v0.assign(tags=lambda df: df.tags.str.lower())[cols].drop_duplicates(),
    final.assign(tags=lambda df: df.tags.str.lower()).rename(col_names, axis=1)[cols].drop_duplicates(),
    on=["sab", "code"],
    how="outer",
    suffixes=("_v0", ""),
    indicator=True
)
print("ONLY in v0", (diff._merge == "left_only").sum())
print("NOT in v0", (diff._merge == "right_only").sum())
#changed_tags = (~pd.isna(diff.tags) | ~pd.isna(diff.tags_v0)) & (diff.tags != diff.tags_v0)
#print("CHANGED TAGS", changed_tags.sum())
diff[diff._merge == "left_only"]

ONLY in v0 13
NOT in v0 438


sab       code     _merge
1553    ICPC2EENG        K05  left_only
1752  SNOMEDCT_US  251120003  left_only
1753  SNOMEDCT_US  251123001  left_only
1754  SNOMEDCT_US  251124007  left_only
1755  SNOMEDCT_US  276796006  left_only
1757  SNOMEDCT_US   29320008  left_only
1758  SNOMEDCT_US  314208002  left_only
1759  SNOMEDCT_US  421869004  left_only
1760  SNOMEDCT_US  425615007  left_only
1761  SNOMEDCT_US  427393009  left_only
1762  SNOMEDCT_US   49710005  left_only
1763  SNOMEDCT_US  733534002  left_only
1764  SNOMEDCT_US   74615001  left_only

In [98]:
dedup_v0.query('dedup_code == "K05"')["sab code str dedup_result dedup_coding_system dedup_code dedup_code_name".split()]

sab code                        str          dedup_result  \
288  ICPC  K05  Irregular heartbeat other  CHANGE_CODING_SYSTEM   

    dedup_coding_system dedup_code            dedup_code_name  
288           ICPC2EENG        K05  Irregular heartbeat other

In [100]:
dedup.query('dedup_code == "K05"')["sab code str dedup_result dedup_sab dedup_code dedup_str".split()]

sab code                        str   dedup_result dedup_sab dedup_code  \
288  ICPC  K05  Irregular heartbeat other  BY_CODE_EQUIV      ICPC        K05   

                           dedup_str  
288  Other abn/irreg heartbeat/pulse

In [71]:
dedup.drop([c for c in dedup.columns if c.startswith('review_')], axis=1).query("code == 'D68.5'")

sab   code                    str cui concept_name     tags comments  \
9     ICD10  D68.5  Primary thrombophilia                   exclude            
76  ICD10DA  D68.5      Primær trombofili                   exclude            
9     ICD10  D68.5  Primary thrombophilia                    narrow            
76  ICD10DA  D68.5      Primær trombofili                    narrow            

            dedup_result dedup_comments dedup_code              dedup_str  \
9   CHANGE_CODING_SYSTEM              -      D68.5  Primary thrombophilia   
76                  NONE              -          -                      -   
9   CHANGE_CODING_SYSTEM              -      D68.5  Primary thrombophilia   
76                  NONE              -          -                      -   

   dedup_sab dedup_cui dedup_ttys dedup_ignore  \
9    ICD10CM  C2584620      AB,HT        false   
76         -         -          -            -   
9    ICD10CM  C2584620      AB,HT        false   
76         -         -          -            -   

                                       file  
9   B_COAGDEF_AESI_Coagulation deficiencies  
76  B_COAGDEF_AESI_Coagulation deficiencies  
9    B_COAGDEF_COV_Coagulation deficiencies  
76   B_COAGDEF_COV_Coagulation deficiencies

In [72]:
dedup_v0.drop([c for c in dedup_v0.columns if c.startswith('review_')], axis=1).query("code == 'D68.5'")

coding_system   code              code_name concept concept_name     tags  \
9          ICD10  D68.5  Primary thrombophilia                       exclude   
76       ICD10DA  D68.5      Primær trombofili                       exclude   
9          ICD10  D68.5  Primary thrombophilia                        narrow   
76       ICD10DA  D68.5      Primær trombofili                        narrow   

   comments          dedup_result dedup_comment dedup_code  ...  \
9            CHANGE_CODING_SYSTEM             -      D68.5  ...   
76                           NONE             -          -  ...   
9            CHANGE_CODING_SYSTEM             -      D68.5  ...   
76                           NONE             -          -  ...   

   dedup_coding_system dedup_concept  \
9              ICD10CM      C2584620   
76                   -             -   
9              ICD10CM      C2584620   
76                   -             -   

                                  dedup_changed dedup_ttys dedup_ignore  \
9   changed coding system from ICD10|cui from -      AB,HT        false   
76                                            -          -            -   
9   changed coding system from ICD10|cui from -      AB,HT        false   
76                                            -          -            -   

   dedup_names_by_code dedup_codes_by_name dedup_original_code  \
9                    -                   -               D68.5   
76                   -                D685               D68.5   
9                    -                   -               D68.5   
76                   -                D685               D68.5   

   dedup_original_code_name                                     file  
9     Primary thrombophilia  B_COAGDEF_AESI_Coagulation deficiencies  
76        Primær trombofili  B_COAGDEF_AESI_Coagulation deficiencies  
9     Primary thrombophilia   B_COAGDEF_COV_Coagulation deficiencies  
76        Primær trombofili   B_COAGDEF_COV_Coagulation deficiencies  

[4 rows x 21 columns]

In [49]:
diff[changed_tags]

sab       code                                       str  \
9         ICD10CM        I66               embolism of cerebral artery   
61        MTHICD9     444.22  Arterial embolism or thrombosis, femoral   
67          ICD10        D66         Hereditary factor VIII deficiency   
68          ICD10        D66         Hereditary factor VIII deficiency   
71          ICD10        D67           Hereditary factor IX deficiency   
...           ...        ...                                       ...   
4285  SNOMEDCT_US  425615007                                       NaN   
4286  SNOMEDCT_US  427393009                                       NaN   
4287  SNOMEDCT_US   49710005                                       NaN   
4288  SNOMEDCT_US  733534002                                       NaN   
4289  SNOMEDCT_US   74615001                                       NaN   

         tags                                             str_v0  tags_v0  \
9      narrow                                                NaN      NaN   
61     narrow                                                NaN      NaN   
67    exclude                  Hereditary factor VIII deficiency   narrow   
68     narrow                  Hereditary factor VIII deficiency  exclude   
71    exclude                    Hereditary factor IX deficiency   narrow   
...       ...                                                ...      ...   
4285      NaN                  Chronic atrial flutter (disorder)   narrow   
4286      NaN      Electrocardiogram: sinus arrhythmia (finding)   narrow   
4287      NaN                       Sinus bradycardia (disorder)   narrow   
4288      NaN  Electrocardiographic complete left bundle bran...   narrow   
4289      NaN                 Tachycardia-bradycardia (disorder)   narrow   

          _merge  
9      left_only  
61     left_only  
67          both  
68          both  
71          both  
...          ...  
4285  right_only  
4286  right_only  
4287  right_only  
4288  right_only  
4289  right_only  

[931 rows x 7 columns]

In [86]:
dedup = load("AESI-dedup").rename(col_names, axis=1)
dedup_v0 = load("VAC4EU-v0/AESI-dedup").rename(col_names, axis=1)
dedup

sab               code  \
0          ICD10                I74   
1          ICD10              I74.0   
2          ICD10              I74.1   
3          ICD10              I74.2   
4          ICD10              I74.3   
..           ...                ...   
787  SNOMEDCT_US          698247007   
788  SNOMEDCT_US          698247007   
789  SNOMEDCT_US          698247007   
790  SNOMEDCT_US          720448006   
791  SNOMEDCT_US  15964901000119107   

                                                   str       cui  \
0                     Arterial embolism and thrombosis  C0155749   
1           Embolism and thrombosis of abdominal aorta             
2    Embolism and thrombosis of other and unspecifi...             
3    Embolism and thrombosis of arteries of upper e...             
4    Embolism and thrombosis of arteries of lower e...             
..                                                 ...       ...   
787                      Other cardiac dysrhythmia NOS             
788                         Other cardiac dysrhythmias             
789                                 Cardiac arrhythmia  C0003811   
790                             Typical atrial flutter  C3264370   
791                            Atypical atrial flutter  C3264372   

                        concept_name    tags comments dedup_result  \
0    Arterial embolus and thrombosis  Narrow                 EXACT   
1                                     Narrow                 EXACT   
2                                     Narrow                 EXACT   
3                                     Narrow                 EXACT   
4                                     Narrow                 EXACT   
..                               ...     ...      ...          ...   
787                                   Narrow               BY_CODE   
788                                   Narrow               BY_CODE   
789               Cardiac Arrhythmia  Narrow                 EXACT   
790           Typical atrial flutter  Narrow                 EXACT   
791          Atypical atrial flutter  Narrow                 EXACT   

    dedup_comments         dedup_code  ... review_content_0 review_author_1  \
0                -                I74  ...              NaN             NaN   
1                -              I74.0  ...              NaN             NaN   
2                -              I74.1  ...              NaN             NaN   
3                -              I74.2  ...              NaN             NaN   
4                -              I74.3  ...              NaN             NaN   
..             ...                ...  ...              ...             ...   
787              -          698247007  ...                     Joan Fortuny   
788              -          698247007  ...                     Joan Fortuny   
789              -          698247007  ...                     Joan Fortuny   
790              -          720448006  ...                     Joan Fortuny   
791              -  15964901000119107  ...                     Joan Fortuny   

           review_date_1 review_content_1 review_author_2  \
0                    NaN              NaN             NaN   
1                    NaN              NaN             NaN   
2                    NaN              NaN             NaN   
3                    NaN              NaN             NaN   
4                    NaN              NaN             NaN   
..                   ...              ...             ...   
787  2022-12-12 00:00:00        no change            sima   
788  2022-12-12 00:00:00        no change            sima   
789  2022-12-12 00:00:00        no change            sima   
790  2022-12-12 00:00:00        no change            sima   
791  2022-12-12 00:00:00        no change            sima   

           review_date_2 review_content_2 review_author_3 review_date_3  \
0                    NaN              NaN             NaN           NaN   
1                    NaN              NaN             